network/xil_eth/xil_eth.c

```diff
@@ -0,0 +1,654 @@
+		       param->mac_ethernet_address, param->emac_baseaddr)) {
+		pr_err("Error adding N/W interface\n\r");
+		ret = FAILURE;
+		goto cleanup;
+	}
+
+	netif_set_default(netif);
+
+	/* now enable interrupts */
+	platform_enable_interrupts();
+
+	if (!param->use_static_ip) {
+		/* specify that the network if is up */
+		netif_set_up(netif);
+
+		/* DHCP Config */
+
+		/* Create a new DHCP client for this interface.
+		* Note: you must call dhcp_fine_tmr() and dhcp_coarse_tmr() at
+		* the predefined regular intervals after starting the client.
+		*/
+		dhcp_start(netif);
+		dhcp_timoutcntr = 24;
+
+		while(((netif->ip_addr.addr) == 0) && (dhcp_timoutcntr > 0))
+			xemacif_input(netif);
+
+		if (dhcp_timoutcntr <= 0) {
+			if ((netif->ip_addr.addr) == 0) {
+				pr_err("DHCP Timeout\r\n");
+				ret = FAILURE;
+				goto cleanup;
+			}
+		}
+	}
+
+	xil_eth_config_noos_if(ldesc);
+
+	*desc = ldesc;
+	for (i = 0; i < MAX_SOCKETS; ++i) {
+		ldesc->sockets[i].id = i;
+		ldesc->sockets[i].desc = ldesc;
+	}
+
+	return SUCCESS;
+cleanup:
+	free(ldesc);
+	return ret;
+}
+
+int32_t xil_eth_remove(struct xil_eth_desc *desc)
+{
+	free(desc);
+
+	return SUCCESS;
+}
+
+/* They seem to be non static in tcp.c but they aren't in any .h file */
+void tcp_fasttmr(void);
+void tcp_slowtmr(void);
+
+/* It is simpler to call tcp_tmr every 250 seconds than using xil_eth_step */
+int32_t xil_eth_step(struct xil_eth_desc *desc)
+{
+	if (TcpFastTmrFlag) {
+		tcp_fasttmr();
+		TcpFastTmrFlag = 0;
+	}
+	if (TcpSlowTmrFlag) {
+		tcp_slowtmr();
+		TcpSlowTmrFlag = 0;
+	}
+	xemacif_input(&desc->netif);
+
+	return SUCCESS;
+}
+
+int32_t xil_eth_get_network_interface(struct xil_eth_desc *desc,
+				      struct network_interface **net)
+{
+	*net = &desc->noos_net;
+
+	return SUCCESS;
+}
+
+int32_t xil_eth_get_ip(struct xil_eth_desc *desc, char *ip_buff,
+		       uint32_t buff_size)
+{
+	ip_addr_t *ip;
+
+	ip = &desc->netif.ip_addr;
+
+	return snprintf(ip_buff, buff_size, "%d.%d.%d.%d", ip4_addr1(ip),
+			ip4_addr2(ip), ip4_addr3(ip), ip4_addr4(ip));
+}
+
+
+/******************************************************************************/
+/*************************** Functions Declarations *******************************/
+/******************************************************************************/
+
+/* Called when new data arrives */
+static err_t xil_eth_recv_callback(void *arg, struct tcp_pcb *tpcb,
+				   struct pbuf *p, err_t err)
+{
+	struct socket_desc *sock = arg;
+
+	if (!p) {
+		/* Remote has closed connection */
+		tcp_recv(sock->pcb, NULL);
+		sock->state = SOCKET_DISCONNECTED;
+
+		return ERR_OK;
+	} else if (err != ERR_OK) {
+		/* Cleanup for unknown reasons */
+		pbuf_free(p);
+		return err;
+	}
+
+	if (!sock->p) {
+		/* First buffer. Save it in p */
+		sock->p = p;
+		sock->p_idx = 0;
+	} else {
+		/*
+		 * TODO: Can this cause error while reading?
+		 * Should some locking be use here?
+		 */
+		/* Concatenate new buffer to p */
+		pbuf_cat(sock->p, p);
```

Comment:
Question: do we need to update `sock->p_idx` here or does `pbuf_cat` update it itself?

Reply:
I assume p is put at the end of the list. 
p_idx is the index in the buffer of the first element